Triangle array

jamwaffles opened the pull request:
Thank you for helping out with embedded-graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

Closes #482.

Should the constructor still take 3 separate arguments? We do have the `Triangle::from_points` method to take arrays and slices, so that might be enough, although maybe it should be renamed?

Benchmark performance hasn't changed.

rfuest commented:
> Should the constructor still take 3 separate arguments?

IMO there is no big advantage in changing it to use an array. Keeping it as is would be one less breaking API change.

> We do have the `Triangle::from_points` method to take arrays and slices, so that might be enough,...

`Triangle::from_points` is kind of odd, because I think it's more generic than any other constructor. And I think that `AsRef` should be used instead of `Borrow`. I would prefer to remove the `Into<Point>` stuff and only accept `AsRef<[Point; 3]>`.

>  ...although maybe it should be renamed?

I think it should. `from_points` can be confusing because `Triangle::points` also exists, which does something unrelated. Should it be called `from_vertices` or maybe `with_vertices`?

jamwaffles commented:
> Triangle::from_points is kind of odd, because I think it's more generic than any other constructor. And I think that AsRef should be used instead of Borrow. I would prefer to remove the Into<Point> stuff and only accept AsRef<[Point; 3]>.

I think it needs to remain `Borrow` if we want to keep accepting both `&[]` and `[]` (which would be nice tbh) for [one reason](https://doc.rust-lang.org/std/convert/trait.AsRef.html):

> Unlike `AsRef`, `Borrow` has a blanket impl for any `T`, and can be used to accept either a reference or a value.

The other changes make sense, so I've removed the `Into<Point>` stuff and renamed it to `with_vertices`, which aligns with `Rectangle::with_corners`, etc.

rfuest commented:
> I think it needs to remain `Borrow` if we want to keep accepting both `&[]` and `[]` (which would be nice tbh) for [one reason](https://doc.rust-lang.org/std/convert/trait.AsRef.html):

OK, I didn't read the docs correctly. I'm just wondering how common this kind of constructor is. Would it be more familiar for users to have separate constructors for arrays and slices? A slice constructor might be more useful than one that only works with references to arrays with 3 elements. Or is there an easy way to convert `&[Point]` into `&[Point; 3]`?

jamwaffles commented:
We could remove the method for now and add a better version in later if someone approaches us with a more specific use case to guide the impl. IIRC it was originally added as a "nice to have", so there's no specific use case for it.

rfuest commented:
I would prefer to at least add a constructor that takes an array. IMO it's strange if the user can access the vertices using an array but cannot construct a new triangle using an array.

Being able to construct a triangle from a slice would be useful for constructing triangles from an iterator like `points.chunks_exact(3)`.

jamwaffles commented:
I could keep the current `from_vertices` to cater for the `[Point; 3]` and `&[Point; 3]` case, but I'm not quite sure how a `from_slice` would behave. It could take the first 3 elements from the slice, but what if the slice is shorter? Return `Err`? Return `None`?

rfuest commented:
Maybe we could implement `core::convert::TryFrom<&[Points]> for Triangle`?

jamwaffles commented:
We certainly could, although is it a better convention to have a `from_slice` method? I've seen that in a few places, but I could be wrong. Either way, it should return a `Result<Self, ()>`, right?

rfuest commented:
I'm not sure how common `from_slice` is. But if it returns a `Result` would it be better to call it `try_from_slice`?

Another option would be to panic in `from_slice` if the slice has the wrong size, which is the behavior some methods in the standard library use: https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice. Or we could provide both `from_slice` and `try_from_slice`. If the user can guarantee that the slice will always have the required size, e.g. in the `chunks_exact` example above, `from_slice` might be easier to use.

jamwaffles commented:
For now, how about we just go with a panicking `from_slice()`. It satisfies the `chunks_exact(3)`, `windows(3)`, etc, use case, and we can add `try_from_slice` at a later date if users want it.

rfuest commented:
Sounds good to me.

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -144,39 +138,37 @@ impl ContainsPoint for Triangle {
 impl Triangle {
-    /// Create a new triangle with a given style
+    /// Create a new triangle with the given vertices.
     pub const fn new(p1: Point, p2: Point, p3: Point) -> Self {
```
This might be better if someone uses code completion to show the arguments of the function:
```suggestion
    pub const fn new(vertex1: Point, vertex2: Point, vertex3: Point) -> Self {
```
Don't worry I don't expect you to change all `p1`...`p3`s in the file :wink:.

jamwaffles replied:
How about `vN`, etc, or is that too ambiguous compared to `vertexN` do you think?

rfuest replied:
Is there a reason why you want to keep the names short? I just prefer more descriptive names in the public API. But I'm also OK with using `vN` if you prefer that.

jamwaffles replied:
No that's a good point actually. I wanted to keep them consistent with `pN` used internally, but didn't consider these args are part of the public API. I've changed them to `vertexN`.

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -144,39 +138,37 @@ impl ContainsPoint for Triangle {
     ///
-    /// This supports both [`Point`]s, as well as anything that implements `Into<Point>` like
-    /// `(i32, i32)`.
+    /// This method will panic if the given slice is less than 3 items long.
```
Perhaps we should panic for all length other than 3. This would match the behavior of `copy_from_slice`.

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -330,14 +313,15 @@ impl Transform for Triangle {
-        self.p1 += by;
-        self.p2 += by;
-        self.p3 += by;
+        self.vertices[0] += by;
+        self.vertices[1] += by;
+        self.vertices[2] += by;
```
```suggestion
        self.vertices.iter_mut().for_each(|v| *v += by);
```

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -185,33 +175,35 @@ impl Triangle {
-        let Self { p1, p2, p3 } = self;
+        let [p1, p2, p3] = self.vertices;
 
         let determinant = -p2.y * p3.x + p1.y * (p3.x - p2.x) + p1.x * (p2.y - p3.y) + p2.x * p3.y;
```
```suggestion
        let determinant = self.area_doubled();
```

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -185,33 +175,35 @@ impl Triangle {
             // Triangle is colinear. Sort points so they lie sequentially along the line.
             Ordering::Equal => {
-                let (p1, p2, p3) = sort_yx(*p1, *p2, *p3);
+                let (p1, p2, p3) = sort_yx(p1, p2, p3);
 
                 Self::new(p1, p2, p3)
```
```suggestion
                self.sorted_yx(p1, p2, p3)
```

After this change `sort_yx` isn't required anymore and you can move the code in `sort_yx` into `sorted_yx`.

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -185,33 +175,35 @@ impl Triangle {
         match determinant.cmp(&0) {
             // Triangle is wound CCW. Swap two points to make it CW.
-            Ordering::Less => Self::new(*p2, *p1, *p3),
+            Ordering::Less => Self::new(p2, p1, p3),
```
After the other changes are applied this is the only place that still uses `p1`...`p3`, so I would suggest to simply change it to this:
```suggestion
            Ordering::Less => Self::new(self.vertices[1], self.vertices[0], self.vertices[2]),
```
Or move the `let [p1, p2, p3] = self.vertices;` into the `Less` match arm, if you prefer that.

rfuest commented on src/primitives/triangle/mod.rs:
```diff
@@ -144,39 +135,38 @@ impl ContainsPoint for Triangle {
-            p1: points[0].into(),
-            p2: points[1].into(),
-            p3: points[2].into(),
+    // MSRV: Consider subslice patterns for rust >= 1.42.0
```
We don't need subslice patterns anymore, now that we require exactly 3 elements.
```suggestion
```

rfuest reviewed (approved):
LGTM

jamwaffles commented:
Thanks!
